src/Tasks/Microsoft.NET.Build.Tasks/ValidateExecutableReferences.cs

```diff
@@ -76,10 +94,14 @@ protected override void ExecuteCore()
                         referencedProjectIsSelfContained = true;
                     }
 
+                    //  We need to check if referenced project will become SelfContained because of its RuntimeIdentifier. This only happens on TargetFrameworks less than net8.0.
+                    var sdkVersionWhereRuntimeIdentifierNoLongerInfersSelfContained = new FrameworkName(".NETCoreApp,Version=v8.0");
+                    bool runtimeIdentifierInfersSelfContained = referencedProjectTargetFramework != null && referencedProjectTargetFramework.Identifier == ".NETCoreApp" && referencedProjectTargetFramework.Version < sdkVersionWhereRuntimeIdentifierNoLongerInfersSelfContained.Version;
+
                     //  If the project is NOT RID agnostic, then a global RuntimeIdentifier will flow to it.
                     //  If the project didn't explicitly specify a value for SelfContained, then this will
                     //  set SelfContained to true
-                    if (runtimeIdentifierIsGlobalProperty && !referencedProjectHadSelfContainedSpecified)
+                    if (RuntimeIdentifierInfersSelfContained && runtimeIdentifierIsGlobalProperty && !referencedProjectHadSelfContainedSpecified)
```

Comment:
I think this needs to refer to the local variable, rather than the property which doesn't ever seem to be set.

```suggestion
                    if (runtimeIdentifierInfersSelfContained && runtimeIdentifierIsGlobalProperty && !referencedProjectHadSelfContainedSpecified)
```